src/util/log.c

```diff
@@ -0,0 +1,648 @@
+        if (mem_fd < 0) {
+                log->error = error_origin(-errno);
+                return false;
+        }
+
+        r = ftruncate(mem_fd, LOG_SIZE_MAX);
+        if (r < 0) {
+                log->error = error_origin(-errno);
+                return false;
+        }
+
+        p = mmap(NULL, LOG_SIZE_MAX, PROT_READ | PROT_WRITE, MAP_SHARED, mem_fd, 0);
+        if (p == MAP_FAILED) {
+                log->error = error_origin(-errno);
+                return false;
+        }
+
+        log->mem_fd = mem_fd;
+        log->map = p;
+        mem_fd = -1;
+        return true;
+}
+
+static int log_fd_send(int destination_fd, int payload_fd) {
+        union {
+                char buffer[CMSG_SPACE(sizeof(int))];
+                struct cmsghdr cmsg;
+        } control;
+        struct msghdr msg;
+        ssize_t l;
+
+        control.cmsg.cmsg_len = CMSG_LEN(sizeof(int));
+        control.cmsg.cmsg_level = SOL_SOCKET;
+        control.cmsg.cmsg_type = SCM_RIGHTS;
+        memcpy(CMSG_DATA(&control.cmsg), &payload_fd, sizeof(int));
+
+        msg = (struct msghdr){
+                .msg_control = &control.cmsg,
+                .msg_controllen = control.cmsg.cmsg_len,
+        };
+
+        l = sendmsg(destination_fd, &msg, MSG_NOSIGNAL);
+        if (l)
+                return (l < 0) ? error_origin(-errno) : error_origin(-ENOTRECOVERABLE);
+
+        return 0;
+}
+
+static int log_loop_send(int destination_fd, const void *blob, size_t n_blob) {
+        ssize_t l;
+
+        while (n_blob > 0) {
+                l = send(destination_fd, blob, n_blob, MSG_NOSIGNAL);
+                if (l < 0)
+                        return error_origin(-errno);
+                else if (l == 0 || l > (ssize_t)n_blob)
+                        return error_origin(-ENOTRECOVERABLE);
+
+                blob += l;
+                n_blob -= l;
+        }
+
+        return 0;
+}
+
+static int log_stream_send(Log *log) {
+        bool log_warn = false;
+        const void *blob = log->map;
+        size_t n_blob = log->offset;
+        ssize_t l;
+        int r, v;
+
+        /*
+         * Stream sockets are a bit nasty since they lack atomic writes. Hence,
+         * whenever we end up with a short-write in lossy mode, we need to
+         * finish the current message in synchronous mode, and then log a
+         * warning. Otherwise, we end up with a garbled output.
+         *
+         * Now, we remember whenever we dropped message. From then on, we only
+         * ever write messages if the output buffer of our socket is empty.
+         * This guarantees the operation will be non-blocking, as long as
+         * suitably sized log messages are used.
+         */
+
+        if (log->lossy) {
+                if (log->n_dropped) {
+                        r = ioctl(log->log_fd, SIOCOUTQ, &v);
+                        if (r < 0) {
+                                return error_origin(-errno);
+                        } else if (v) {
```

Comment:
Shouldn't we be checking not for the queue being empty, but if the message will fit. Otherwise you have 1) dropping messages unnecessarily and 2) writing out partial messages that are too big to fit even in an empty queue (and hence blocking later in the function).

Reply:
How?

Reply:
Just add a comment to explain why it is not possible (if you agree that that's what we would have liked to do). Also, isn't 2) a problem? What's stopping someone from repeatedly triggering overly long log messages that will end up blocking every time the output queue is empty?

Reply:
How can you trigger overly long log messages? This is the stderr-backend, it uses just the MESSAGE= part. That should always fit into the kernel buffer. If not, you're doing something wrong, aren't you?

Reply:
Ah, missed that part. That's fine then. Comment would be nice.